invoker/core/src/main/java/com/google/cloud/functions/invoker/TypedFunctionExecutor.java

```diff
@@ -0,0 +1,195 @@
+        .filter(method -> method.getName().equals(APPLY_METHOD) && method.getParameterCount() == 1)
+        .map(method -> method.getGenericParameterTypes()[0])
+        .findFirst();
+  }
+
+  /** Executes the user's method, can handle all HTTP type methods. */
+  @Override
+  public void service(HttpServletRequest req, HttpServletResponse res) {
+    HttpRequestImpl reqImpl = new HttpRequestImpl(req);
+    HttpResponseImpl respImpl = new HttpResponseImpl(res);
+    ClassLoader oldContextClassLoader = Thread.currentThread().getContextClassLoader();
+
+    try {
+      Thread.currentThread().setContextClassLoader(func.getClass().getClassLoader());
+      handleRequest(reqImpl, respImpl);
+    } finally {
+      Thread.currentThread().setContextClassLoader(oldContextClassLoader);
+      flushResponse(respImpl);
+    }
+  }
+
+  private void handleRequest(HttpRequest req, HttpResponse resp) {
+    WireFormat format = configuration.getFormat();
+
+    Object reqObj = null;
+    try {
+      reqObj = format.deserialize(req, argType);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to parse request for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Object respObj = null;
+    try {
+      respObj = this.func.apply(reqObj);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to execute " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+
+    try {
+      format.serialize(respObj, resp);
+    } catch (Throwable t) {
+      logger.log(Level.SEVERE, "Failed to serialize response for " + func.getClass().getName(), t);
+      resp.setStatusCode(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return;
+    }
+  }
+
+  private void flushResponse(HttpResponseImpl respImpl) {
+    try {
+      // We can't use HttpServletResponse.flushBuffer() because we wrap the
+      // PrintWriter returned by HttpServletResponse in our own BufferedWriter
+      // to match our API. So we have to flush whichever of getWriter() or
+      // getOutputStream() works.
+      try {
+        respImpl.getOutputStream().flush();
+      } catch (IllegalStateException e) {
+        respImpl.getWriter().flush();
+      }
+    } catch (IOException e) {
+      // Too bad, can't flush.
+    }
+  }
+
+  private static class ConfigurationImpl implements TypedFunction.Configuration {
+    private WireFormat format = null;
+
+    WireFormat getFormat() {
+      if (format == null) {
+        synchronized (this) {
+          if (format == null) {
+            format = new GsonWireFormat();
+          }
+        }
+      }
+
+      return format;
+    }
+
+    @Override
+    public Configuration setWireFormat(WireFormat format) {
+      this.format = format;
+      return null;
+    }
+  }
+
+  private static class GsonWireFormat implements TypedFunction.WireFormat {
```

Comment:
Do these two nested interfaces need to exist if they only have one Impl each? Or when and where will the other implementations be?

Reply:
These interfaces decouple our `functions-framework-api` package from our implementations in the `invoker` package (this is the bit that gets injected by the Cloud Function buildpacks).

`WireFormat` needs to be an interface because alternative implementations for Protobuf decoding/encoding may be provided in the future.

`Configuration` needs to be an interface because our `ConfigurationImpl` provide a GSON wire format as a default implementation and we don't want a dependency on `Gson` in the `function-framework-api` package.